### PR TITLE
Add filemanager for rendering content from gotemplates

### PR DESCRIPTION
**Pull request by corest**
Towards https://github.com/giantswarm/giantswarm/issues/2816

As the current ignition spec uses data schema uri for files, all the files will be moved into separate templates and then rendered into base64 representation. 

E.g. `v_4_0_0` will have folder `files` with subfolders `config`, `kubeconfig`, `k8s-resources` etc. Main master cloudconfig then will use next construction for rendering file in operating system

```
    - path: /srv/psp_policies.yaml
      filesystem: root
      mode: 0644
      contents:
        source: "data:text/plain;charset=utf-8;base64,{{  index .Files "k8s-resource/psp_policies.yaml" }}"
```

### CODE REVIEW

File: v_4_0_0/filemanager.go [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+	"path"
+	"text/template"
+
+	"github.com/giantswarm/microerror"
+)
+
+// Files is map[string]string for files that we fetched from disk and then filled with data.
+type Files map[string]string
+
+func RenderFiles(filespath string, ctx interface{}) (*Files, error) {
```
Comment (kopiczko) [inline v_4_0_0/filemanager.go]: Why there is a pointer returned? In general map should be returned as a value.

File: v_4_0_0/filemanager.go [hunk 2]
```diff
@@ -0,0 +1,42 @@
...
+	files := Files{}
+	dirList, err := ioutil.ReadDir(filespath)
+	if err != nil {
+		return nil, microerror.Maskf(err, "Failed to read files dir: %s, error: %#v", filespath, err)
+	}
+
+	for _, dir := range dirList {
+		fileList, err := ioutil.ReadDir(path.Join(filespath, dir.Name()))
+		if err != nil {
+			return nil, microerror.Maskf(err, "Failed to read dir: %s, error: %#v", path.Join(filespath, dir.Name()), err)
```
Comment (kopiczko) [inline v_4_0_0/filemanager.go]: Error annotations should be lower-cased, i.e. `failed to read...`.

File: v_4_0_0/filemanager.go (hunk 1 above)
Comment (kopiczko) [reply v_4_0_0/filemanager.go]: `filespath` is supposed to be a directory. I would make that clear in the name.

File: v_4_0_0/filemanager.go [hunk 3]
```diff
@@ -0,0 +1,42 @@
...
+			tmpl, err := template.ParseFiles(path.Join(filespath, dir.Name(), file.Name()))
+			if err != nil {
+				return nil, microerror.Maskf(err, "Failed to file: %s, error: %#v", path.Join(filespath, dir.Name(), file.Name()), err)
+			}
+			var data bytes.Buffer
+			tmpl.Execute(&data, ctx)
+
+			files[dir.Name()+"/"+file.Name()] = base64.StdEncoding.EncodeToString(data.Bytes())
+		}
+	}
```
Comment (kopiczko) [inline v_4_0_0/filemanager.go]: I'm not sure what do you intend to use this function for but [filepath.Walk](https://golang.org/pkg/path/filepath/#Walk) may be a better candidate.

File: v_4_0_0/filemanager.go [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+
+		for _, file := range fileList {
+			tmpl, err := template.ParseFiles(path.Join(filespath, dir.Name(), file.Name()))
+			if err != nil {
+				return nil, microerror.Maskf(err, "Failed to file: %s, error: %#v", path.Join(filespath, dir.Name(), file.Name()), err)
+			}
+			var data bytes.Buffer
+			tmpl.Execute(&data, ctx)
+
+			files[dir.Name()+"/"+file.Name()] = base64.StdEncoding.EncodeToString(data.Bytes())
```
Comment (tuommaki) [inline v_4_0_0/filemanager.go]: Maybe use `path.Join()` instead of `dir.Name()+"/"+file.Name()`?

**Comment by corest** [comment]
refactored `RenderFiles`, ptal

File: v_4_0_0/filemanager.go [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+	"path/filepath"
+	"text/template"
+
+	"github.com/giantswarm/microerror"
+)
+
+// Files is map[string]string for files that we fetched from disk and then filled with data.
+type Files map[string]string
+
+func RenderFiles(filesdir string, ctx interface{}) (Files, error) {
```
Comment (tuommaki) [inline v_4_0_0/filemanager.go]: Maybe add godoc here? For example:
```go
// RenderFiles walks over filesdir and parses all regular files with
// text/template. Parsed templates are then rendered with ctx, base64 encoded
// and added to returned Files.                                                 
//
// filesdir must not contain any other files than templates that can be parsed
// with text/template.
```

File: v_4_0_0/filemanager.go [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+	"encoding/base64"
+	"fmt"
+	"os"
+	"path/filepath"
+	"text/template"
+
+	"github.com/giantswarm/microerror"
+)
+
+// Files is map[string]string for files that we fetched from disk and then filled with data.
```
Comment (tuommaki) [inline v_4_0_0/filemanager.go]: Would it be worth to add mention about key/val meaning:
```go
// Files is map[string]string (k: filename, v: contents) for files taht we fetched from disk and then filled with data.
```

File: v_4_0_0/filemanager.go [hunk 7]
```diff
@@ -0,0 +1,43 @@
...
+		if f.Mode().IsRegular() {
+			tmpl, err := template.ParseFiles(path)
+			if err != nil {
+				return microerror.Maskf(err, "failed to parse file #%q", path)
+			}
+			var data bytes.Buffer
+			tmpl.Execute(&data, ctx)
+
+			relativePath, err := filepath.Rel(filesdir, path)
+			fmt.Printf("path: %s", relativePath)
```
Comment (tuommaki) [inline v_4_0_0/filemanager.go]: Is this print intentional for final use or possible leftover from debugging?

Comment (corest) [reply v_4_0_0/filemanager.go]: leftover

File: v_4_0_0/filemanager.go [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+// filesdir must not contain any other files than templates that can be parsed
+// with text/template.
+func RenderFiles(filesdir string, ctx interface{}) (Files, error) {
+	files := Files{}
+
+	err := filepath.Walk(filesdir, func(path string, f os.FileInfo, err error) error {
+		if f.Mode().IsRegular() {
+			tmpl, err := template.ParseFiles(path)
+			if err != nil {
+				return microerror.Maskf(err, "failed to parse file #%q", path)
```
Comment (kopiczko) [inline v_4_0_0/filemanager.go]: It's `%#q` :)

**Pull request merged**